### ISSUE TITLE
vmap(fori_loop(...)) should not batch the loop counter

### ISSUE THREAD

**Issue by hawkinsp** [google/jax#1190]
This led to a significant performance problem in the `threefry2x32` implementation when it was changed to use a `dynamic_slice` as part of its implementation and when batched as part of the `gamma` implementation. If we cannot prove that `i` is equal across batch elements, `x[i]` turns into a relatively expensive gather of scalars, but if we know `x` and `i` are unbatched, it remains a dynamic slice.

**Comment by mattjj** [comment]
Is this a duplicate of #1188 ? 

By the way, we have a much improved vmap of while_loop/fori_loop waiting in the no-more-tuples branch. That branch needs at least another couple days of work though...

**Comment by hawkinsp** [comment]
I don't know. We should try it once that branch lands; maybe we can close both issues at the same time.

**Comment by mattjj** [comment]
We fixed this in #1224 (or soon thereafter):

```python
In [10]: make_jaxpr(vmap(lambda x: lax.fori_loop(0, 4, lambda _, x: x + 1, x)))(np.arange(3))
Out[10]:
{ lambda  ;  ; a.
  let b c d = while[ cond_nconsts=0
                     cond_jaxpr={ lambda  ;  ; a b c.
                                  let d = lt a b
                                  in [d] }
                     body_nconsts=0
                     body_jaxpr={ lambda  ;  ; a b c.
                                  let d = add a 1
                                      e = add c 1
                                  in [d, b, e] } ] 0 4 a
  in [d] }
```

Notice that the `cond_jaxpr` just operates on a scalar (you have to do some mental shape inference) and doesn't have a `reduce_or` in it.
